src/runtime/tracing.cpp

```diff
@@ -10,7 +10,105 @@ typedef int32_t (*trace_fn)(void *, const halide_trace_event_t *);
 
 namespace Halide { namespace Runtime { namespace Internal {
 
-WEAK int halide_trace_file = 0;
+// A spinlock that allows for shared and exclusive access. It's
+// equivalent to a reader-writer lock, but in my case the "readers"
+// will actually be writing simultaneously to the trace buffer, so
+// that's a bad name. We use the __sync primitives used elsewhere in
+// the runtime for atomic work. They are well supported by clang.
+class SharedExclusiveSpinLock {
+    uint32_t lock;
+    const static uint32_t exclusive_held_mask = 0x80000000;
+    const static uint32_t exclusive_waiting_mask = 0x40000000;
+    const static uint32_t shared_mask = 0x3fffffff;
```

Comment:
Please document as "Count of threads currently holding shared access. Must be zero if the exclusive bit is set. Cannot increase if the waiting bit is set."